File: netDxf/Objects/DxfLayerIndex.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace netDxf.Objects
{
    /// <summary>One stored layer name and its hard-owned IDBUFFER in a LAYER_INDEX.</summary>
    /// <remarks>The name is stored text, not a reference to the layer table. Buffer counts are derived from the current IDBUFFER sequence.</remarks>
    public sealed class DxfLayerIndexEntry
    {
        /// <summary>Creates an entry for a nonempty stored layer name and an IDBUFFER.</summary>
        public DxfLayerIndexEntry(string layerName, DxfIdBuffer buffer)
        {
            if (string.IsNullOrEmpty(layerName) || layerName.IndexOfAny(new[] { '\0', '\r', '\n' }) >= 0)
                throw new ArgumentException("A layer-index name must be nonempty single-line Unicode text.", nameof(layerName));
            for (int i = 0; i < layerName.Length; i++)
            {
                if (char.IsHighSurrogate(layerName[i]))
                {
                    if (i + 1 >= layerName.Length || !char.IsLowSurrogate(layerName[++i]))
                        throw new ArgumentException("A layer-index name contains an unpaired surrogate.", nameof(layerName));
                }
                else if (char.IsLowSurrogate(layerName[i]))
                    throw new ArgumentException("A layer-index name contains an unpaired surrogate.", nameof(layerName));
            }
            this.LayerName = layerName;
            this.Buffer = buffer ?? throw new ArgumentNullException(nameof(buffer));
        }
        /// <summary>Gets the exact stored name, including case and unresolved names.</summary>
        public string LayerName { get; }
        /// <summary>Gets the IDBUFFER hard-owned by this entry's index.</summary>
        public DxfIdBuffer Buffer { get; }
        /// <summary>Gets the group-90 count derived from all buffer entries, including duplicates and nulls.</summary>
        public int Count { get { return this.Buffer.References.Count; } }
    }

    /// <summary>A stored LAYER_INDEX timestamp and ordered layer-name/IDBUFFER entries.</summary>
    /// <remarks>This object preserves a stored index graph; it does not build, update, or evaluate a layer index. Layer-table renaming does not rewrite stored names.</remarks>
    public sealed class DxfLayerIndex : DxfDatabaseObject
    {
        private double timestamp;
        private List<DxfLayerIndexEntry> entries = new List<DxfLayerIndexEntry>();
        /// <summary>Creates a detached index with no entries and a zero timestamp.</summary>
        public DxfLayerIndex() : base("LAYER_INDEX") { }
        /// <summary>Gets or sets the finite group-40 Julian-date value without calendar conversion.</summary>
        public double Timestamp
        {
            get { return this.timestamp; }
            set
            {
                if (double.IsNaN(value) || double.IsInfinity(value)) throw new ArgumentOutOfRangeException(nameof(value), "The stored timestamp must be finite.");
                this.timestamp = value;
            }
        }
        /// <summary>Gets the ordered entries; duplicate names are retained.</summary>
        public IReadOnlyList<DxfLayerIndexEntry> Entries { get { return this.entries.AsReadOnly(); } }
        /// <summary>Atomically replaces entries and establishes reciprocal ownership of their IDBUFFER objects.</summary>
        /// <remarks>Detached indexes can adopt unowned detached buffers. Registered indexes can rename or reorder entries using the same owned buffer set; adding or orphaning registered children is rejected. Buffer reference lists remain editable.</remarks>
        public void SetEntries(IEnumerable<DxfLayerIndexEntry> values) { this.SetEntries(values, false); }
        internal void LoadEntries(IEnumerable<DxfLayerIndexEntry> values) { this.SetEntries(values, true); }
        private void SetEntries(IEnumerable<DxfLayerIndexEntry> values, bool loading)
        {
            if (values == null) throw new ArgumentNullException(nameof(values));
            var snapshot = values.ToList();
            if (this.IsErased) throw new InvalidOperationException("An erased index cannot adopt objects.");
            var children = new HashSet<DxfIdBuffer>();
            foreach (DxfLayerIndexEntry entry in snapshot)
            {
                if (entry == null) throw new ArgumentException("A layer-index entry cannot be null.", nameof(values));
                DxfIdBuffer child = entry.Buffer;
                if (!children.Add(child)) throw new ArgumentException("Each layer-index entry must own a distinct IDBUFFER.", nameof(values));
                if (child.IsErased) throw new InvalidOperationException("An erased buffer cannot be attached again.");
                if (child.Owner != null && !ReferenceEquals(child.Owner, this)) throw new ArgumentException("An IDBUFFER already has another owner.", nameof(values));
                if (child.Database != this.Database) throw new ArgumentException("An index and its buffers must share their registration state and database.", nameof(values));
                if (DxfObjectDatabase.IsAncestor(child, this)) throw new ArgumentException("Layer-index ownership cannot form a cycle.", nameof(values));
                if (this.Database != null)
                {
                    this.Database.CheckRegistered(this); this.Database.CheckRegistered(child);
                    if (!ReferenceEquals(child.Owner, this)) throw new ArgumentException("A registered buffer must already be owned by its index.", nameof(values));
                }
            }
            if (this.Database != null && !loading && !children.SetEquals(this.entries.Select(entry => entry.Buffer)))
                throw new InvalidOperationException("Replacing registered entries must preserve their complete owned buffer set.");
            foreach (DxfLayerIndexEntry entry in this.entries)
                if (!children.Contains(entry.Buffer)) entry.Buffer.Owner = null;
            foreach (DxfIdBuffer child in children) child.Owner = this;
            this.entries = snapshot;
        }
        internal override IEnumerable<DxfDatabaseObject> DeclaredOwnedObjects { get { return this.entries.Select(entry => entry.Buffer); } }
        internal override IEnumerable<DxfObject> DatabaseReferences { get { return this.entries.Select(entry => entry.Buffer); } }
        internal override DxfDatabaseObject CloneShell() { return new DxfLayerIndex { Timestamp = this.Timestamp }; }
        internal override void CopyDatabaseReferencesTo(DxfDatabaseObject clone, Func<DxfObject, DxfObject> resolve)
        {
            ((DxfLayerIndex)clone).SetEntries(this.entries.Select(entry => new DxfLayerIndexEntry(entry.LayerName, (DxfIdBuffer)resolve(entry.Buffer))));
        }
        internal override void ValidateDatabaseSchema(DxfObjectDatabase database, List<string> errors)
        {
            if (this.Database != null && !(this.Owner is DxfDictionary)) errors.Add("LAYER_INDEX requires a dictionary owner.");
            if (this.entries.Any(entry => !ReferenceEquals(entry.Buffer.Owner, this))) errors.Add("LAYER_INDEX ownership is not reciprocal.");
            if (this.entries.Select(entry => entry.Buffer).Distinct().Count() != this.entries.Count) errors.Add("LAYER_INDEX repeats an owned IDBUFFER.");
        }
    }
}

File: javascript/netDxf/Objects/DxfLayerIndex.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
import { DxfDatabaseObject, DxfDictionary } from './DxfDatabaseObject.js';
import { CheckStoredName } from '../../runtime/CheckedCollection.js';
import { ReferenceList } from '../../runtime/ReferenceList.js';
import { IsAncestor, ReadOnlyReferenceView } from '../../runtime/DatabaseModel.js';
import { ArgumentException, ArgumentNullException, ArgumentOutOfRangeException, InvalidOperationException } from '../../runtime/Errors.js';
export class DxfLayerIndexEntry {
  constructor(layerName, buffer) {
    CheckStoredName(layerName, 'layerName');
    if (buffer == null) throw new ArgumentNullException('buffer');
    Object.defineProperties(this, { LayerName: { value: layerName, enumerable: true }, Buffer: { value: buffer, enumerable: true } });
    Object.freeze(this);
  }
  get Count() { return this.Buffer.References.Count; }
}
export class DxfLayerIndex extends DxfDatabaseObject {
  #timestamp = 0; #entries = new ReferenceList();
  constructor() { super('LAYER_INDEX'); }
  get Timestamp() { return this.#timestamp; }
  set Timestamp(value) { if (!Number.isFinite(value)) throw new ArgumentOutOfRangeException('value'); this.#timestamp = value; }
  get Entries() { return ReadOnlyReferenceView(this.#entries); }
  SetEntries(values) { this.#setEntries(values, false); }
  LoadEntries(values) { this.#setEntries(values, true); }
  #setEntries(values, loading) {
    if (values == null) throw new ArgumentNullException('values');
    const snapshot = Array.from(values);
    if (this.IsErased) throw new InvalidOperationException('An erased index cannot adopt objects.');
    const children = new Set();
    for (const entry of snapshot) {
      if (entry == null) throw new ArgumentException('A layer-index entry cannot be null.', 'values');
      const child = entry.Buffer;
      if (children.has(child)) throw new ArgumentException('Each entry must own a distinct IDBUFFER.', 'values');
      children.add(child);
      if (child.IsErased) throw new InvalidOperationException('An erased buffer cannot be attached again.');
      if (child.Owner != null && child.Owner !== this) throw new ArgumentException('An IDBUFFER already has another owner.', 'values');
      if (child.Database !== this.Database) throw new ArgumentException('Index and buffers must share a database and registration state.', 'values');
      if (IsAncestor(child, this)) throw new ArgumentException('Layer-index ownership cannot form a cycle.', 'values');
      if (this.Database != null) {
        this.Database.CheckRegistered(this); this.Database.CheckRegistered(child);
        if (child.Owner !== this) throw new ArgumentException('A registered buffer must already belong to this index.', 'values');
      }
    }
    if (this.Database != null && !loading) {
      const before = new Set(Array.from(this.#entries, entry => entry.Buffer));
      if (children.size !== before.size || [...children].some(child => !before.has(child)))
        throw new InvalidOperationException('Registered entries must preserve their complete owned buffer set.');
    }
    for (const entry of this.#entries) if (!children.has(entry.Buffer)) entry.Buffer.Owner = null;
    for (const child of children) child.Owner = this;
    this.#entries = new ReferenceList(snapshot);
  }
  get DeclaredOwnedObjects() { return Array.from(this.#entries, entry => entry.Buffer); }
  get DatabaseReferences() { return this.DeclaredOwnedObjects; }
  CloneShell() { const copy = new DxfLayerIndex(); copy.Timestamp = this.Timestamp; return copy; }
  CopyDatabaseReferencesTo(clone, resolve) {
    clone.SetEntries(Array.from(this.#entries, entry => new DxfLayerIndexEntry(entry.LayerName, resolve(entry.Buffer))));
  }
  ValidateDatabaseSchema(database, errors) {
    if (this.Database != null && !(this.Owner instanceof DxfDictionary)) errors.Add('LAYER_INDEX requires a dictionary owner.');
    if ([...this.#entries].some(entry => entry.Buffer.Owner !== this)) errors.Add('LAYER_INDEX ownership is not reciprocal.');
    if (new Set(Array.from(this.#entries, entry => entry.Buffer)).size !== this.#entries.Count) errors.Add('LAYER_INDEX repeats an owned IDBUFFER.');
  }
}
